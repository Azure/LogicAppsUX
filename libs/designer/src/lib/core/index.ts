export * from './DesignerProvider';
export * from './BJSWorkflowProvider';
export * from './ProviderWrappedContext';
export { getReactQueryClient } from './ReactQueryProvider';
export type { RootState, AppDispatch } from './store';
export { store } from './store';
export { templateStore } from './state/templates/store';
export {
  useConnectionMapping,
  useConnectionRefs,
  useIsOperationMissingConnection,
} from './state/connection/connectionSelector';
export type { NodeInputs } from './state/operation/operationMetadataSlice';
export {
  useOperationsInputParameters,
  useNodesInitialized,
  useNodesAndDynamicDataInitialized,
} from './state/operation/operationSelector';
export type { ErrorMessage } from './state/workflow/workflowInterfaces';
export {
  discardAllChanges,
  setFocusNode,
  setIsWorkflowDirty,
  setHostErrorMessages,
} from './state/workflow/workflowSlice';
export {
  useIsWorkflowDirty,
  useNodeDisplayName,
  useNodeMetadata,
} from './state/workflow/workflowSelectors';
export {
  useIsWorkflowParametersDirty,
  useWorkflowParameterValidationErrors,
} from './state/workflowparameters/workflowparametersselector';
export { useIsDesignerDirty, resetDesignerDirtyState } from './state/global';
export { useAllSettingsValidationErrors } from './state/setting/settingSelector';
export { useAllConnectionErrors } from './state/operation/operationSelector';
export { serializeWorkflow, parseWorkflowParameterValue } from './actions/bjsworkflow/serializer';
export {
  setSelectedNodeId,
  changePanelNode,
  clearPanel,
  openPanel,
  collapsePanel,
} from './state/panel/panelSlice';
export { useOperationInfo } from './state/selectors/actionMetadataSelector';
export { useReplacedIds } from './state/workflow/workflowSelectors';
export {
  useConnectionPanelSelectedNodeIds,
  useDiscoveryPanelSelectedNodeIds,
  useOperationPanelPinnedNodeId,
  useOperationPanelSelectedNodeId,
} from './state/panel/panelSelectors';
export { initializeServices } from './state/designerOptions/designerOptionsSlice';
export { resetWorkflowState, resetNodesLoadStatus } from './state/global';
export { TemplatesDataProvider } from './templates/TemplatesDataProvider';
export { TemplatesDesignerProvider } from './templates/TemplatesDesignerProvider';
export {
  validateParameter,
  getCustomCodeFilesWithData,
} from './utils/parameters/helper';
export {
  createLiteralValueSegment,
  createTokenValueSegment,
  ValueSegmentConvertor,
  type ValueSegmentConvertorOptions,
} from './utils/parameters/segment';
export {
  getOutputTokenSections,
  getExpressionTokenSections,
} from './utils/tokens';
export { getTriggerNodeId } from './utils/graph';
export { updateParameterValidation } from './state/operation/operationMetadataSlice';
export { updateWorkflowParameters } from './actions/bjsworkflow/initialize';
export {
  getBrandColorFromManifest,
  getIconUriFromManifest,
} from './utils/card';
export { addOperation } from './actions/bjsworkflow/add';
export { updateNodeConnection } from './actions/bjsworkflow/connections';
export { storeStateToUndoRedoHistory, onUndoClick, onRedoClick } from './actions/bjsworkflow/undoRedo';
export { useCanUndo, useCanRedo } from './state/undoRedo/undoRedoSelectors';
