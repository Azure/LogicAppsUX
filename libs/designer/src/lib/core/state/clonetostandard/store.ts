import { combineReducers, configureStore } from '@reduxjs/toolkit';
import resourceReducer from './resourceslice';
import cloneReducer from './cloneslice';
import cloneOptionsReducer from './cloneoptionsslice';

const rootReducer = combineReducers({
  resource: resourceReducer,
  clone: cloneReducer,
  cloneOptions: cloneOptionsReducer,
});

export const setupStore = (preloadedState?: Partial<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
};

export const cloneStandardStore = setupStore();
export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
