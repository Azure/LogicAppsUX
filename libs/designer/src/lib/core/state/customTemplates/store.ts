import { combineReducers, configureStore } from '@reduxjs/toolkit';
import type {} from 'redux-thunk';
import workflowReducer from './workflowSlice';
import templateReducer from './templateSlice';
import customTemplateOptionsReducer from './customTemplateOptionsSlice';
import panelReducer from './panelSlice';

const rootReducer = combineReducers({
  workflow: workflowReducer,
  template: templateReducer,
  customTemplateOptions: customTemplateOptionsReducer,
  panel: panelReducer,
});

export const setupStore = (preloadedState?: Partial<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
};

export const templateStore = setupStore();
export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
