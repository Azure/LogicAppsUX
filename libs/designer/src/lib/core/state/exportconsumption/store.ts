import { combineReducers, configureStore } from '@reduxjs/toolkit';
import resourceReducer from './resourceslice';

const rootReducer = combineReducers({
  resource: resourceReducer,
});

export const setupStore = (preloadedState?: Partial<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
};

export const exportConsumptionStore = setupStore();
export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
