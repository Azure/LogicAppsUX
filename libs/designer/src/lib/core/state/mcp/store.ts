import { combineReducers, configureStore } from '@reduxjs/toolkit';
import mcpOptionsReducer from './mcpOptions/mcpOptionsSlice';
import operationReducer from '../operation/operationMetadataSlice';
import panelReducer from './panel/mcpPanelSlice';
import connectionReducer from '../connection/connectionSlice';
import resourceReducer from './resourceSlice';
import mcpSelectionReducer from './mcpselectionslice';

const rootReducer = combineReducers({
  resource: resourceReducer,
  operations: operationReducer,
  connection: connectionReducer,
  mcpOptions: mcpOptionsReducer,
  mcpPanel: panelReducer,
  mcpSelection: mcpSelectionReducer,
});

export const setupStore = (preloadedState?: Partial<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
};

export const mcpStore = setupStore();
export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
