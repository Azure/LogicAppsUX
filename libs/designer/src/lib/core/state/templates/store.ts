import { combineReducers, configureStore } from '@reduxjs/toolkit';
import type {} from 'redux-thunk';
import workflowReducer from './workflowSlice';
import templateReducer from './templateSlice';
import manifestReducer from './manifestSlice';
import panelReducer from './panelSlice';

const rootReducer = combineReducers({
  workflow: workflowReducer,
  template: templateReducer,
  manifest: manifestReducer,
  panel: panelReducer,
});

export const setupStore = (preloadedState?: Partial<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
};

export const templateStore = setupStore();
export const templatesPathFromState = '../../templates/templateFiles';
export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
