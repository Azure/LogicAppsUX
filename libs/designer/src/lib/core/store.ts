import connectionsReducer from './state/connection/connectionSlice';
import designerOptionsReducer from './state/designerOptions/designerOptionsSlice';
import designerViewReducer from './state/designerView/designerViewSlice';
import operationMetadataReducer from './state/operation/operationMetadataSlice';
import panelReducer from './state/panel/panelSlice';
import settingsReducer from './state/setting/settingSlice';
import staticResultsSchemasReducer from './state/staticresultschema/staticresultsSlice';
import tokens from './state/tokensSlice';
import workflowReducer from './state/workflow/workflowSlice';
import workflowParametersReducer from './state/workflowparameters/workflowparametersSlice';
import { configureStore } from '@reduxjs/toolkit';

export const store = configureStore({
  reducer: {
    workflow: workflowReducer,
    operations: operationMetadataReducer,
    panel: panelReducer,
    connections: connectionsReducer,
    settings: settingsReducer,
    designerOptions: designerOptionsReducer,
    designerView: designerViewReducer,
    tokens: tokens,
    workflowParameters: workflowParametersReducer,
    staticResults: staticResultsSchemasReducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: false,
    }),
});

if (process.env.NODE_ENV === 'development') {
  (window as any).DesignerStore = store;
}

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
