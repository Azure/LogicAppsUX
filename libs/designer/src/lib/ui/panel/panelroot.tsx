import constants from '../../common/constants';
import { useMonitoringView, useReadOnly } from '../../core/state/designerOptions/designerOptionsSelectors';
import {
  useIsDiscovery,
  useIsPanelCollapsed,
  useRegisteredPanelTabs,
  useSelectedNodeId,
  useSelectedPanelTabName,
  useVisiblePanelTabs,
} from '../../core/state/panel/panelSelectors';
import {
  collapsePanel,
  expandPanel,
  isolateTab,
  registerPanelTabs,
  selectPanelTab,
  setTabVisibility,
  showDefaultTabs,
} from '../../core/state/panel/panelSlice';
import { useIconUri, useOperationInfo } from '../../core/state/selectors/actionMetadataSelector';
import { useNodeDescription, useNodeMetadata } from '../../core/state/workflow/workflowSelectors';
import { setNodeDescription } from '../../core/state/workflow/workflowSlice';
import { aboutTab } from './panelTabs/aboutTab';
import { codeViewTab } from './panelTabs/codeViewTab';
import { createConnectionTab } from './panelTabs/createConnectionTab';
import { monitoringTab } from './panelTabs/monitoringTab';
import { parametersTab } from './panelTabs/parametersTab';
import { scratchTab } from './panelTabs/scratchTab';
import { selectConnectionTab } from './panelTabs/selectConnectionTab';
import { SettingsTab } from './panelTabs/settingsTab';
import { RecommendationPanelContext } from './recommendation/recommendationPanelContext';
import { isNullOrUndefined, SUBGRAPH_TYPES } from '@microsoft-logic-apps/utils';
import type { MenuItemOption, PageActionTelemetryData } from '@microsoft/designer-ui';
import { MenuItemType, PanelContainer, PanelHeaderControlType, PanelLocation, PanelScope, PanelSize } from '@microsoft/designer-ui';
import React, { useCallback, useEffect, useState } from 'react';
import { useIntl } from 'react-intl';
import { useDispatch } from 'react-redux';

export const PanelRoot = (): JSX.Element => {
  const intl = useIntl();
  const dispatch = useDispatch();

  const readOnly = useReadOnly();
  const isMonitoringView = useMonitoringView();

  const collapsed = useIsPanelCollapsed();
  const selectedNode = useSelectedNodeId();
  const isDiscovery = useIsDiscovery();

  const [width, setWidth] = useState(PanelSize.Auto);

  const registeredTabs = useRegisteredPanelTabs();
  const visibleTabs = useVisiblePanelTabs();
  const selectedPanelTab = useSelectedPanelTabName();

  const comment = useNodeDescription(selectedNode);
  const operationInfo = useOperationInfo(selectedNode);
  const iconUriResult = useIconUri(operationInfo);
  const nodeMetaData = useNodeMetadata(selectedNode);
  let showCommentBox = !isNullOrUndefined(comment);

  useEffect(() => {
    const tabs = [monitoringTab, parametersTab, aboutTab, codeViewTab, SettingsTab, scratchTab, createConnectionTab, selectConnectionTab];
    if (process.env.NODE_ENV !== 'production') {
      tabs.push(scratchTab);
    }
    dispatch(registerPanelTabs(tabs));
  }, [dispatch]);

  useEffect(() => {
    dispatch(
      setTabVisibility({
        tabName: constants.PANEL_TAB_NAMES.MONITORING,
        visible: isMonitoringView,
      })
    );
  }, [dispatch, isMonitoringView]);

  useEffect(() => {
    if (!visibleTabs?.map((tab) => tab.name.toLowerCase())?.includes(selectedPanelTab ?? ''))
      dispatch(selectPanelTab(visibleTabs[0]?.name.toLowerCase()));
  }, [dispatch, visibleTabs, selectedPanelTab]);

  const setSelectedTab = (tabName: string | undefined) => {
    dispatch(selectPanelTab(tabName));
  };

  useEffect(() => {
    if (nodeMetaData && nodeMetaData.subgraphType === SUBGRAPH_TYPES.SWITCH_CASE) {
      dispatch(isolateTab(constants.PANEL_TAB_NAMES.PARAMETERS));
    } else {
      dispatch(showDefaultTabs());
    }
    dispatch(
      setTabVisibility({
        tabName: constants.PANEL_TAB_NAMES.MONITORING,
        visible: isMonitoringView,
      })
    );
  }, [dispatch, selectedNode, nodeMetaData, isMonitoringView]);

  useEffect(() => {
    dispatch(
      setTabVisibility({
        tabName: constants.PANEL_TAB_NAMES.PARAMETERS,
        visible: isMonitoringView || operationInfo?.type !== 'Scope',
      })
    );
  }, [dispatch, isMonitoringView, operationInfo]);

  useEffect(() => {
    collapsed ? setWidth(PanelSize.Auto) : setWidth(PanelSize.Medium);
  }, [collapsed]);

  const collapse = useCallback(() => {
    dispatch(collapsePanel());
  }, [dispatch]);

  const expand = useCallback(() => {
    dispatch(expandPanel());
  }, [dispatch]);

  const getPanelHeaderControlType = (): boolean => {
    // TODO: 13067650
    return isDiscovery;
  };

  const getPanelHeaderMenu = (): MenuItemOption[] => {
    const menuOptions: MenuItemOption[] = [];
    // TODO: 13067650 Conditionals to decide when to show these menu options
    getCommentMenuItem(menuOptions);
    getDeleteMenuItem(menuOptions);
    return menuOptions;
  };

  const getCommentMenuItem = (options: MenuItemOption[]): MenuItemOption[] => {
    const commentDescription = intl.formatMessage({
      defaultMessage: 'Comment',
      description: 'Comment text',
    });
    const disabledCommentAction = intl.formatMessage({
      defaultMessage: 'Comments can only be added while editing the inputs of a step.',
      description: 'Text to tell users why a comment is disabled',
    });
    const commentAdd = intl.formatMessage({
      defaultMessage: 'Add a comment',
      description: 'Text to tell users to click to add comments',
    });
    const commentDelete = intl.formatMessage({
      defaultMessage: 'Delete comment',
      description: 'Text to tell users to click to delete comments',
    });

    options.push({
      disabled: readOnly,
      type: MenuItemType.Advanced,
      disabledReason: disabledCommentAction,
      iconName: 'Comment',
      key: commentDescription,
      title: showCommentBox ? commentDelete : commentAdd,
      onClick: handleCommentMenuClick,
    });
    return options;
  };

  const getDeleteMenuItem = (options: MenuItemOption[]): MenuItemOption[] => {
    const deleteDescription = intl.formatMessage({
      defaultMessage: 'Delete',
      description: 'Delete text',
    });
    // TODO: 13067650 Disabled reason/description tobe implemented when panel actions gets built
    const canDelete = true;
    const disabledDeleteAction = intl.formatMessage({
      defaultMessage: 'This operation has already been deleted.',
      description: 'Text to tell users why delete is disabled',
    });

    options.push({
      key: deleteDescription,
      disabled: readOnly || !canDelete,
      disabledReason: disabledDeleteAction,
      iconName: 'Delete',
      title: deleteDescription,
      type: MenuItemType.Advanced,
      onClick: handleDelete,
    });
    return options;
  };

  const handleCommentMenuClick = (_: React.MouseEvent<HTMLElement>): void => {
    showCommentBox = !showCommentBox;
    dispatch(setNodeDescription({ nodeId: selectedNode, ...(showCommentBox && { description: '' }) }));
  };

  // TODO: 12798945? onClick for delete when node store gets built
  const handleDelete = (): void => {
    // TODO: 12798935 Analytics (event logging)
  };

  const togglePanel = (): void => (!collapsed ? collapse() : expand());

  return isDiscovery ? (
    <RecommendationPanelContext isCollapsed={collapsed} toggleCollapse={togglePanel} width={width} />
  ) : (
    <PanelContainer
      cardIcon={iconUriResult.result}
      comment={comment}
      panelLocation={PanelLocation.Right}
      isCollapsed={collapsed}
      noNodeSelected={!selectedNode}
      isLoading={iconUriResult.isLoading}
      panelScope={PanelScope.CardLevel}
      panelHeaderControlType={getPanelHeaderControlType() ? PanelHeaderControlType.DISMISS_BUTTON : PanelHeaderControlType.MENU}
      panelHeaderMenu={getPanelHeaderMenu()}
      selectedTab={selectedPanelTab}
      showCommentBox={showCommentBox}
      tabs={registeredTabs}
      width={width}
      onDismissButtonClicked={handleDelete}
      readOnlyMode={readOnly}
      setSelectedTab={setSelectedTab}
      toggleCollapse={togglePanel}
      trackEvent={handleTrackEvent}
      onCommentChange={(value) => {
        dispatch(setNodeDescription({ nodeId: selectedNode, description: value }));
      }}
      title={selectedNode}
    />
  );
};

// TODO: 12798935 Analytics (event logging)
// eslint-disable-next-line @typescript-eslint/no-empty-function
const handleTrackEvent = (_data: PageActionTelemetryData): void => {};
