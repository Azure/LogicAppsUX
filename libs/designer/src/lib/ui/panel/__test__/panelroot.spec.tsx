import { store } from '../../../core/store';
import type { PanelRootProps } from '../panelroot';
import { PanelRoot } from '../panelroot';
import * as React from 'react';
import { Provider } from 'react-redux';
import * as ReactShallowRenderer from 'react-test-renderer/shallow';

describe('ui/workflowparameters/workflowparameter', () => {
  let minimal: PanelRootProps, renderer: ReactShallowRenderer.ShallowRenderer;

  beforeEach(() => {
    minimal = {};
    renderer = ReactShallowRenderer.createRenderer();
  });

  afterEach(() => {
    renderer.unmount();
  });

  it('should construct.', () => {
    const panel = renderer.render(
      <Provider store={store}>
        {' '}
        <PanelRoot {...minimal} />{' '}
      </Provider>
    );
    expect(panel).toMatchSnapshot();
  });
});
