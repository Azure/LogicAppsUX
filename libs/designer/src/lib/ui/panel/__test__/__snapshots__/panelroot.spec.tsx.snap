// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/workflowparameters/workflowparameter should construct. 1`] = `
<Context.Provider
  value={
    Object {
      "getServerState": undefined,
      "store": Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
   
  <PanelRoot />
   
</Context.Provider>
`;
